pipelines/templates/api-deploy-environment.yml

```diff
@@ -0,0 +1,94 @@
+      - task: Bash@3
+        name: createMatrixJson
+        displayName: Create Matrix
+        inputs:
+          targetType: inline
+          script: |
+            echo $(regionList)
+            source ./deployment/arm/utils.sh
+            matrix=$(toRegionMatrix $(regionList))
+            echo $matrix
+            echo "##vso[task.setVariable variable=matrix;isOutput=true]$matrix"
+
+  - job: RegionalMaxtrixDeployment
+    displayName: "Deploy Region:"
+    dependsOn: setup
+    strategy:
+      matrix: $[ dependencies.setup.outputs['createMatrixJson.matrix'] ]
+    steps:
+      - task: UseDotNet@2
+        displayName: Dotnet version 3.1.100
+        inputs:
+          version: '3.1.100'
+
+      - task: DownloadBuildArtifacts@0
+        displayName: Download Build Artifacts
+        inputs:
+          buildType: 'current'
+          downloadType: 'single'
+          artifactName: 'drop'
+          downloadPath: '$(Build.ArtifactStagingDirectory)'
+
+      - task: AzureRMWebAppDeployment@4
+        displayName: Deploy to Staging Slot
+        inputs:
+          ConnectionType: 'AzureRM'
+          azureSubscription: ${{ parameters.subscription }}
+          appType: 'apiApp'
+          ResourceGroupName: 'rg-${{ parameters.businessUnit }}-${{ parameters.appName }}-${{ parameters.environment }}-$(region)'
+          WebAppName: 'backend-${{ parameters.businessUnit }}-${{ parameters.appName }}-${{ parameters.environment }}-$(region)'
+          deployToSlotOrASE: true
+          SlotName: staging
+          packageForLinux: '$(Build.ArtifactStagingDirectory)/**/*.zip'  
+          appSettings: 
+            -CosmosSettings:ServiceEndpoint $(COSMOSDB_ENDPOINT) 
+            -CosmosSettings:AuthKey $(COSMOSDB_AUTHKEY) 
+            -CosmosSettings:DatabaseName linkylinkdb 
+            -CosmosSettings:ContainerName linkbundles
+            -AzureAdB2C:Instance $(AZUREADB2C_INSTANCE)
+            -AzureAdB2C:ClientId $(AZUREADB2C_CLIENTID)
+            -AzureAdB2C:Domain $(AZUREADB2C_DOMAIN)
+            -AzureAdB2C:SignUpSignInPolicyId $(AZUREADB2C_POLICYID)
+
+      - task: DotNetCoreCLI@2
+        displayName: Validation Tests
+        inputs:
+          command: test
+          projects: '**/LinkyLink.Integration.Tests/*.csproj'
+          arguments: -c $(BuildConfiguration)
+        env:
+          INTTEST_AzureAdB2C__BaseAddress: 'https://backend-${{ parameters.businessUnit }}-${{ parameters.appName }}-${{ parameters.environment }}-$(region)-staging.azurewebsites.net'
+      
+      - task: AzureAppServiceManage@0
+        displayName: Swap Staging and Production Slots
+        inputs:
+          azureSubscription: ${{ parameters.subscription }}
+          action: Swap Slots
+          ResourceGroupName: 'rg-${{ parameters.businessUnit }}-${{ parameters.appName }}-${{ parameters.environment }}-$(region)'
+          WebAppName: 'backend-${{ parameters.businessUnit }}-${{ parameters.appName }}-${{ parameters.environment }}-$(region)'
```

Comment:
This is nit-picking, but since we are utilizing our naming conventions from here, can we add a comment to the relevant script files to mention that any changes to the names should update the pipelines as well? And vice-versa if possible?

Reply:
+1.  I would suggest moving the naming conventions to a util script (that other scripts also reference) that we run in a task before this.  Said task would export those value as environment variables and this task can just reference the values.  

Having the naming convention logic in multiple places is a setup for error later on when things change.

Reply:
I didn't quite follow this comment. Could you elaborate?

Reply:
have a task before this that run a script that generate the ResourceGroupName, WebAppName etc. and export them as env variable.

Then in this task, instead of 

```ResourceGroupName: 'rg-${{ parameters.businessUnit }}-${{ parameters.appName }}-${{ parameters.environment }}-$(region)'```

we'll have
```ResourceGroupName: $(rgName)```

Similar to how Wallace create the matrix value in another section